src/beast/include/beast/impl/basic_streambuf.ipp

```diff
@@ -478,74 +478,77 @@ auto
         if(n > avail)
         {
+            out_end_ = out_->size();
             n -= avail;
-            out_end_ = pos->size();
-            while(++pos != list_.end())
-            {
-                if(n < pos->size())
-                {
-                    out_end_ = n;
-                    n = 0;
-                    ++pos;
-                    break;
-                }
-                n -= pos->size();
-            }
         }
         else
         {
-            ++pos;
             out_end_ = out_pos_ + n;
             n = 0;
         }
+        debug_check();
     }
-
-    if(n > 0)
+    while(n > 0 && ! reuse.empty())
     {
-        assert(pos == list_.end());
-        for(;;)
+        auto& e = reuse.front();
+        reuse.erase(reuse.iterator_to(e));
+        list_.push_back(e);
+        if(n > e.size())
         {
-            auto const size = std::max(alloc_size_, n);
-            auto& e = *reinterpret_cast<element*>(
-                alloc_traits::allocate(this->member(),
-                    size + sizeof(element)));
-            alloc_traits::construct(this->member(), &e, size);
-            list_.push_back(e);
-            if(out_ == list_.end())
-            {
-                out_ = list_.iterator_to(e);
-                debug_check();
-            }
-            if(n <= size)
-            {
-                out_end_ = n;
-                debug_check();
-                break;
-            }
-            n -= size;
-            debug_check();
+            out_end_ = e.size();
+            n -= e.size();
+        }
+        else
+        {
+            out_end_ = n;
+            n = 0;
         }
+        debug_check();
     }
-    else
+    while(n > 0)
     {
-        while(pos != list_.end())
+        auto const size = std::max(alloc_size_, n);
+        auto& e = *reinterpret_cast<element*>(
+            alloc_traits::allocate(this->member(),
+                sizeof(element) + size));
+        alloc_traits::construct(this->member(), &e, size);
+        list_.push_back(e);
+        if(out_ == list_.end())
+            out_ = list_.iterator_to(e);
+        if(n > e.size())
         {
-            auto& e = *pos++;
-            list_.erase(list_.iterator_to(e));
-            auto const len = e.size() + sizeof(e);
-            alloc_traits::destroy(this->member(), &e);
-            alloc_traits::deallocate(this->member(),
-                reinterpret_cast<std::uint8_t*>(&e), len);
+            out_end_ = e.size();
+            n -= e.size();
+        }
+        else
+        {
+            out_end_ = n;
+            n = 0;
         }
         debug_check();
     }
-
+    for(auto it = reuse.begin(); it != reuse.end();)
+    {
+        auto& e = *it++;
+        reuse.erase(list_.iterator_to(e));
+        auto const len = e.size() + sizeof(e);
+        alloc_traits::destroy(this->member(), &e);
+        alloc_traits::deallocate(this->member(),
+            reinterpret_cast<std::uint8_t*>(&e), len);
+    }
```

Comment:
I'm noting that we carefully move un-needed elements into the local `reuse` list.  And then, here in this final `for` loop, we nuke the contents of `reuse`.  I'm wondering if there might be a better name for `reuse`, since none of what goes in `reuse` is reused?  I suspect the name is a hold-over from a prior implementation.